dra-evolution/pkg/api/quota_types.go

```diff
@@ -0,0 +1,46 @@
+package api
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// Quota controls whether a ResourceClaim may get allocated.
+// Quota is namespaced and applies to claims within the same namespace.
+type Quota struct {
+	metav1.TypeMeta
+	// Standard object metadata.
+	metav1.ObjectMeta
+
+	Spec QuotaSpec
+}
+
+type QuotaSpec struct {
+	// AllowManagementAccess controls whether claims with ManagementAccess
+	// may be allocated. The default if unset is to deny such access.
+	// +optional
+	AllowManagementAccess *bool
+
+	// Other useful future extensions (>= 1.32):
+
+	// DeviceLimits is a CEL expression which take the currently allocated
+	// devices and their attributes and some new allocations as input and
+	// returns false if those allocations together are not permitted in the
+	// namespace.
+	//
+	// DeviceLimits string
+
+	// A class listed in ResourceClassDenyList must not be used in this
+	// namespace. This can be useful for classes which contain
+	// configuration pieces that a user in this namespace should not have
+	// access to.
+	//
+	// ResourceClassDenyList []string
+
+	// A class listed in ResourceClassAllowList may be used in this namespace
+	// even when that class is marked as "privileged". Normally classes
+	// are not privileged and using them does not require explicit listing
+	// here, but some classes may contain more sensitive configurations
+	// that not every user should have access to.
+	//
+	// ResourceClassAllowList []string
```

Comment:
We can use the same mechanism as this class allow/deny list also for individual devices. Here's how:
- Each `Device` in a ResourcePool can optionally be marked as "privileged".
  This gets decided by the vendor, possibly based on some configuration
  of the driver that the cluster admin can modify. If a driver doesn't offer
  this, then a cluster admin can use a mutating webhook to control this
  field.
- DeviceDenyList contains device selectors. If a device instance matches, it
  cannot be used inside the namespace.
- DeviceAllowList is the same, except that if it matches, "privileged" devices may be used.

I think we should include this in 1.31 and also make sure that we can extend this API later. Schedulers must refuse to allocate a claim if they don't fully understand the Quota objects in a namespace, because a future extension might limit whether a claim can be allocated. The current API as proposed doesn't achieve that.

Question:
- Rename Quota -> ResourcePolicy? Not everything in it is a "quota", some is more like "access control". Adding "Resource" ensures a potential shorthand for this API resource type ("resourcepolicies") matches the Go type name.
- Repeat multiple "rules" in a single object or a single "rule" per object? We should support multiple object per namespace, just as with `v1.ResourceQuota`, so we already have repetition through that.
- Add a "spec"? Probably yes, because this is a specification provided by the admin and adding a status (later!) seems plausible.

Reply:
I had considered other approaches for marking a device as "privileged", but separate configuration objects don't "fail safe": a scheduler might not have seen them yet and incorrectly not apply that separate configuration.

Reply:
I'm confused a little bit because I thought class was now optional, so I don't see how it could be used for quota.

Also, I haven't seen a requirement for the concept of a "privileged device".

I do know that admins will want to control which devices can be used by various namespaces. I would think that a CEL expression plus a count would be sufficient in 1.31, plus maybe mgmt access? When we add allocatable device resources (e.g., memory on a device), we might need something that puts aggregate limits on those in a namespace.

Reply:
> I'm confused a little bit because I thought class was now optional, so I don't see how it could be used for quota.

This class allow/deny is for classes with privileged configuration. 

> Also, I haven't seen a requirement for the concept of a "privileged device".

@klueska mentioned that he was doing access control in classic DRA depending on the device type. But I might misremember or might have misunderstood. @klueska, can you clarify?

> I do know that admins will want to control which devices can be used by various namespaces. I would think that a CEL expression plus a count would be sufficient in 1.31, plus maybe mgmt access?

That looks like a reasonable simplification and can replace the "privileged device" feature. One difference is that we default to "allow" (right?) unless someone adds a restriction in each namespace. Are we okay with that?

> When we add allocatable device resources (e.g., memory on a device), we might need something that puts aggregate limits on those in a namespace.

This also applies without allocatable device resources. If "memory" is a normal attribute, then it might make sense to put a limit on the sum of that attribute to "encourage" users to use smaller devices if those are sufficient.

Reply:
Yeah. I am leaning though towards keeping attributes as static simple comparison items, *without any semantic meaning*. The semantic value ("I can aggregate this, it is memory") comes in with the allocatable device resources concept. Just to keep the behaviors simpler to manage.

I guess if the quota/policy kind is embedding the idea that something can be aggregated and that is user added, it's safe to do what you suggest.

Reply:
Any attribute which is a quantity can already be summed up. Whether that is meaningful and useful depends on the semantic, which isn't something that we need to decide - we just need to provide the mechanism to do it.

As I write this, I realize that this is similar to `matchAttributes` vs. "general match CEL expression". We could do the simple cases first with:
- sum of quantity X < maxX
- total number of devices which match device selector < n

This would be easier to use than some (probably hideous) CEL expression and easier to implement.

Reply:
yes, those two would probably cover the fast majority of the cases

Reply:
Answering [@thockin's question](https://github.com/kubernetes-sigs/wg-device-management/pull/14#pullrequestreview-2095042210) here:
> If I write a claim which my quota does not allow (e.g. I ask for 50 GPUs and my quota is 5), does the claim fail to create or fail to schedule?

The claim gets created, but allocating it fails. Perhaps the user is only temporarily over quota and deallocating some other claim will allow the blocked one to proceed.

Basically quota is turned into additional requirements for a request, as if the user had listed them in their claim. 

> Will the autoscaler know not to provision machines with 50 GPUs?

Yes, because provisioning such a machine would not make some unschedulable pod schedulable.

> If I write a OneOf claim, where one of the 2 options would be allowed under my quota, and the other would not - is that allowed?  

That is allowed. The quota then ensures that the allowed alternative in the OneOf is picked.

>  What if the autoscaler was instructed to prefer creating machines with the variety that is over quota?

The autoscaler should not create new machines unless there is at least one unschedulable pod which then becomes schedulable.

It might not be "the best" machine to bring online, but I think that is part of a larger problem with Cluster Autoscaler. I've heard that it doesn't support heterogeneous clusters well.